linker/linker_phdr.cpp

```diff
@@ -608,16 +610,60 @@ bool ElfReader::ReserveAddressSpace(const android_dlextinfo* extinfo) {
   return true;
 }
 
-bool ElfReader::LoadSegments() {
+bool ElfReader::LoadSegments(const android_dlextinfo* extinfo) {
   for (size_t i = 0; i < phdr_num_; ++i) {
     const ElfW(Phdr)* phdr = &phdr_table_[i];
 
     if (phdr->p_type != PT_LOAD) {
       continue;
     }
 
+    bool random_start = false;
+#if defined(__aarch64__) || defined(__arm__)
+    // Randomly map PF_RAND_ADDR segments, but only if the client is not
+    // overriding with a fixed load address
+    random_start = (phdr->p_flags & PF_RAND_ADDR) != 0 &&
+      !((extinfo != nullptr && (
+           extinfo->flags & ANDROID_DLEXT_RESERVED_ADDRESS ||
+           extinfo->flags & ANDROID_DLEXT_FORCE_FIXED_VADDR ||
+           extinfo->flags & ANDROID_DLEXT_LOAD_AT_FIXED_ADDRESS)));
+
+    ElfW(Addr) random_start_address = 0;
+    // Linux provides 8 bits of entropy for mmaps (see
+    // arch/arm/mm/mmap.c). However, after the address space is somewhat full,
+    // this results in little to no actual randomness. We try to fix this here.
+    if (random_start) {
+#if defined(__arm__)
+      // Tentatively using 0xb0000000-0xb6000000 as the pagerando range
+      unsigned long low  = 0xb0000000;
+      unsigned long high = 0xb6000000;
+#elif defined(__aarch64__)
+      // Tentatively using 0x1000000000-0x5000000000 as the pagerando range
+      unsigned long low  = 0x1000000000;
+      unsigned long high = 0x5000000000;
+#endif
```

Comment:
Would do architecture-guarded `#define`s so the actual algorithm can just use those/minimize the `#if-elif-endif` in the code.